### PR TITLE
neorados: Update exerciser/demo to use C++20 coroutines

### PR DESCRIPTION
**Pull request by adamemerson**
Just to work the bugs out and give some demo code.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/tools/neorados.cc [hunk 1]
```diff
@@ -367,10 +374,24 @@ int main(int argc, char* argv[])
...
+		       // generated by `co_spawn`, so we have to catch
+		       // here and propagate the exit value, as well
+		       // as having the larger catch for any
+		       // exceptions thrown by `program_options` or
+		       // elsewhere.
+		       fmt::print(std::cerr, "{}: {}: {}\n", prog, command,
+				  e.what());
+		       exitval = EXIT_FAILURE;
+		     }
+		   }, ba::detached);
```
Comment (cbodley) [inline src/tools/neorados.cc:394]: according to https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/co_spawn/overload1.html, that last `ba::detached` argument could instead be a `handler(std::exception_ptr)`. i would expect that completion handler to run outside of the coroutine, so you could rethrow the exception there. `void std::rethrow_exception(std::exception_ptr p)` has that same signature, can you use it directly?
```c++
      ba::co_spawn(c,
		   [&]() -> ba::awaitable<void> {
		     auto r = co_await R::RADOS::Builder{}.build(
		         c, ba::use_awaitable);
		     co_await ci->f(r, parameters);
		   }, std::rethrow_exception);
```

Comment (adamemerson) [reply src/tools/neorados.cc:394]: That works perfectly, thank you.

**Comment by adamemerson** [comment]
jenkins test make check

File: src/include/neorados/RADOS.hpp [hunk 2]
```diff
@@ -127,9 +127,8 @@ class IOContext final {
 public:
 
   IOContext();
-  explicit IOContext(std::int64_t pool);
-  IOContext(std::int64_t _pool, std::string_view _ns);
-  IOContext(std::int64_t _pool, std::string&& _ns);
+  IOContext(std::int64_t pool);
```
Comment (yuvalif) [inline src/include/neorados/RADOS.hpp:130]: why remove the explicit call when ctor accept one int parameter?

Comment (yuvalif) [reply src/include/neorados/RADOS.hpp:130]: but "pool" is just ant int. this will create implicit conversion from int to IOContext.
isn't thank risky?

**Comment by yuvalif** [comment]
isn't the neorados code currently used with the RGW?
shouldn't the neorados library support both a spawn/yield API and c++20 coroutines API?

**Comment by cbodley** [comment]
> isn't the neorados code currently used with the RGW? shouldn't the neorados library support both a spawn/yield API and c++20 coroutines API?

it does support both as part of asio's [Asynchronous Model](https://think-async.com/Asio/asio-1.22.1/doc/asio/overview/model.html), specifically through our handling of the [CompletionToken](https://think-async.com/Asio/asio-1.22.1/doc/asio/overview/model/completion_tokens.html) template type

File: src/include/neorados/RADOS.hpp (hunk 2 above)
Comment (adamemerson) [reply src/include/neorados/RADOS.hpp:130]: I don't think so, unless I'm missing something. We don't really have any overloads it's likely to catch by accident, and conceptually IOContext (really I should have named it IOLocator so people wouldn't be tempted to stuff operation flags into it or something. Also that's what it maps to in Objector) is a pool (indicated by an int) plus some other less used stuff.

For something that's "Int plus optional other stuff" being able to implicitly accept an int doesn't seem wrong.

Comment (yuvalif) [reply src/include/neorados/RADOS.hpp:130]: i've seen such a bug in our code when we had a non-explicit ctor of `rgw_user` that accepts an int.
my concern is that we have:
```
foo(IOContext ctx, ...)
```
and someone incorrectly invokes it with an int (which is not the pool id) as the first parameter.
the compiler won't complain and implicitly construct an `IOContext` based on an integer that is not a pool id (this is very likely, if `foo` accepts multiple integer parameters, which is quite common).
making the ctor explicit, prevents these kinds of bugs.

**Comment by adamemerson** [comment]
This looks mostly reasonable, I think? I'm not as familiar with how much background failure RBD has.

https://pulpito.ceph.com/aemerson-2022-09-20_03:59:09-rbd-wip-coro-test-distro-default-smithi/

**Comment by yuvalif** [comment]
do we have teuthology tests running `ceph_test_neorados`?

**Comment by adamemerson** [comment]
> do we have teuthology tests running `ceph_test_neorados`?

No, not yet.

File: src/include/neorados/RADOS.hpp (hunk 2 above)
Comment (adamemerson) [reply src/include/neorados/RADOS.hpp:130]: All right, played around with this and the ergonomics of changing it aren't too bad in practice.

Comment (adamemerson) [reply src/include/neorados/RADOS.hpp]: All right. Made this change and tested it out.

**Review by cbodley — APPROVED** [review]

**Review by yuvalif — APPROVED** [review]

**Comment by cbodley** [comment]
> This looks mostly reasonable, I think? I'm not as familiar with how much background failure RBD has.
> 
> https://pulpito.ceph.com/aemerson-2022-09-20_03:59:09-rbd-wip-coro-test-distro-default-smithi/

yikes, from those `cli_generic.sh` failures i see that `rbd bench` is causing osds to report a bunch of slow requests:
```
2022-09-20T06:18:21.142 INFO:tasks.ceph.osd.0.smithi171.stderr:2022-09-20T06:18:21.139+0000 7f9299790700 -1 osd.0 158 get_health_metrics reporting 256 slow ops, oldest is osd_op(client.10593.0:2055 3.3 3:db356461:test1::rbd_data.295e2a328d04.00000000000000f0:head [set-alloc-hint object_size 4194304 write_size 4194304,write 3420160~4096 [fadvise_random] in=4096b] snapc 0=[] RETRY=4 ondisk+retry+write+known_if_redirected+supports_pool_eio e158)
2022-09-20T06:18:21.448 INFO:tasks.ceph.osd.1.smithi171.stderr:2022-09-20T06:18:21.443+0000 7f52fc777700 -1 osd.1 158 get_health_metrics reporting 256 slow ops, oldest is osd_op(client.10593.0:2007 3.2 3:50ad0aa8:test1::rbd_data.295e2a328d04.0000000000000060:head [set-alloc-hint object_size 4194304 write_size 4194304,write 1519616~4096 [fadvise_random] in=4096b] snapc 0=[] RETRY=3 ondisk+retry+write+known_if_redirected+supports_pool_eio e158)
```
do these failures look familiar @idryomov?

**Comment by cbodley** [comment]
@adamemerson i've been watching other runs of the rbd suite like https://pulpito.ceph.com/dis-2022-10-05_10:59:41-rbd-wip-dis-testing-distro-default-smithi/ and haven't seen any of those cli_generic.sh failures. you might try a rebase and rerun to see if they persist?

**Comment by cbodley** [comment]
@idryomov a friendly ping to review the rbd suite results

**Comment by yuvalif** [comment]
@adamemerson @cbodley  any update on this PR?

**Comment by yuvalif** [comment]
@adamemerson ping

**Comment by yuvalif** [comment]
@cbodley there is no input from RBD team on this PR for more than a month. any way to move forward here?

**Comment by adamemerson** [comment]
https://pulpito.ceph.com/aemerson-2022-11-30_20:07:48-rbd-wip-coro-test-distro-default-smithi/

**Comment by cbodley** [comment]
> https://pulpito.ceph.com/aemerson-2022-11-30_20:07:48-rbd-wip-coro-test-distro-default-smithi/

lots of 'test -f /home/ubuntu/cephtest/archive/qemu/client.0/success' failures which also show up in [Yuri's recent baseline](https://pulpito.ceph.com/yuriw-2022-12-09_22:35:49-rbd-main-distro-default-smithi/)

i tried to schedule a --rerun but the shaman packages have disappeared

**Comment by idryomov** [comment]
> > https://pulpito.ceph.com/aemerson-2022-11-30_20:07:48-rbd-wip-coro-test-distro-default-smithi/
> 
> lots of 'test -f /home/ubuntu/cephtest/archive/qemu/client.0/success' failures which also show up in [Yuri's recent baseline](https://pulpito.ceph.com/yuriw-2022-12-09_22:35:49-rbd-main-distro-default-smithi/)
> 
> i tried to schedule a --rerun but the shaman packages have disappeared

Unfortunately this is expected right now -- I'll take this through the rbd suite myself.

**Comment by idryomov** [comment]
There is a trivial conflict in src/librbd/io/ObjectRequest.cc (parent_overlap -> raw_overlap), please rebase.

**Comment by cbodley** [comment]
thanks @idryomov, i pushed a rebase

**Comment by idryomov** [comment]
The lab is broken right now, I'll pick this up in my next rbd integration run.

File: src/include/neorados/RADOS.hpp [hunk 3]
```diff
@@ -745,82 +719,43 @@ class RADOS final
...
+       bl = std::move(bl), this](auto&& handler) mutable {
+	notify_ack(std::move(o), std::move(ioc), std::move(cookie),
+		   std::move(bl), SimpleOpComp::create(get_executor(),
+						       std::move(handler)));
+      }, token);
   }
 
   template<typename CompletionToken>
-  auto unwatch(uint64_t cookie, const IOContext& ioc,
+  auto unwatch(std::uint64_t cookie, IOContext ioc,
```
Comment (idryomov) [inline src/include/neorados/RADOS.hpp:751]: FWIW I don't think sprinkling `std` on trivial integer types is worthwhile.  There is plenty of bare uint64_t uses in this file even after this PR.

**Comment by idryomov** [comment]
Also, the current version doesn't build on Ubuntu Focal and CentOS 8:
```
[ 48%] Building CXX object src/tools/CMakeFiles/neorados.dir/neorados.cc.o
In file included from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/format.h:48,
                 from /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/tools/neorados.cc:31:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h: In instantiation of ‘constexpr fmt::v9::detail::value<Context> fmt::v9::detail::make_value(T&&) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; T = const neorados::IOContext&]’:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:1777:29:   required from ‘constexpr fmt::v9::detail::value<Context> fmt::v9::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; fmt::v9::detail::type <anonymous> = fmt::v9::detail::type::custom_type; T = const neorados::IOContext&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]’
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:1901:77:   required from ‘constexpr fmt::v9::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {const neorados::Object&, const neorados::IOContext&}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Args = {neorados::Object, neorados::IOContext}]’
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:1918:31:   required from ‘constexpr fmt::v9::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<Args>::type>::type ...> fmt::v9::make_format_args(Args&& ...) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Args = {const neorados::Object&, const neorados::IOContext&}]’
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:3206:44:   required from ‘std::string fmt::v9::format(fmt::v9::format_string<T ...>, T&& ...) [with T = {const neorados::Object&, const neorados::IOContext&}; std::string = std::__cxx11::basic_string<char>; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, const neorados::Object&, const neorados::IOContext&>]’
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/tools/neorados.cc:218:13:   required from here
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:1757:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1757 |       formattable,
      |       ^~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-1592-g0937c0c9/rpm/el8/BUILD/ceph-18.0.0-1592-g0937c0c9/src/fmt/include/fmt/core.h:1757:7: note: ‘formattable’ evaluates to false
make[2]: *** [src/tools/CMakeFiles/neorados.dir/build.make:76: src/tools/CMakeFiles/neorados.dir/neorados.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:8955: src/tools/CMakeFiles/neorados.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by adamemerson** [comment]
jenkins test api

**Comment by adamemerson** [comment]
jenkins test api

**Comment by yuvalif** [comment]
@adamemerson any status update?

**Comment by cbodley** [comment]
@adamemerson i know that you've been carrying these commits in your wip-coro-after-reef branch. now that reef has forked, i think we should start by getting just this PR reviewed/tested/merged. i don't want to ask other components to review *everything* in https://github.com/ceph/ceph/pull/49737. but merging this PR could unblock all of the rgw-specific in your feature branch. what do you think?

**Comment by cbodley** [comment]
i also see that the rbd suite is looking [very green for reef](https://pulpito.ceph.com/yuriw-2023-03-03_18:25:49-rbd-reef-distro-default-smithi/), so that should make it easier to validate the results

**Comment by adamemerson** [comment]
> @adamemerson i know that you've been carrying these commits in your wip-coro-after-reef branch. now that reef has forked, i think we should start by getting just this PR reviewed/tested/merged. i don't want to ask other components to review _everything_ in #49737. but merging this PR could unblock all of the rgw-specific in your feature branch. what do you think?

That sounds fine to me, I think.

File: src/include/neorados/RADOS.hpp [hunk 4]
```diff
@@ -1147,4 +1019,15 @@ struct hash<neorados::IOContext> {
...
+namespace fmt {
+template <>
+struct formatter<neorados::Object> : ostream_formatter {};
+
+template <>
+struct formatter<neorados::IOContext> : ostream_formatter {};
+
+template <>
+struct formatter<neorados::Op> : ostream_formatter {};
+} // namespace fmt
```
Comment (cbodley) [inline src/include/neorados/RADOS.hpp:1031]: this appears to be the cause of [centos9 build failures in shaman](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/68579/consoleFull):
```
src/neorados/RADOS.cc:38:
src/include/neorados/RADOS.hpp:1024:56: error: expected class-name before ‘{’ token
 1024 | struct formatter<neorados::Object> : ostream_formatter {};
      |                                                        ^
src/include/neorados/RADOS.hpp:1027:59: error: expected class-name before ‘{’ token
 1027 | struct formatter<neorados::IOContext> : ostream_formatter {};
      |                                                           ^
src/include/neorados/RADOS.hpp:1030:52: error: expected class-name before ‘{’ token
 1030 | struct formatter<neorados::Op> : ostream_formatter {};
      |                                                    ^
```

**Comment by cbodley** [comment]
passed the rgw suite in https://pulpito.ceph.com/cbodley-2023-03-11_01:08:10-rgw-wip-coro-test-distro-default-smithi/

the rbd suite in https://pulpito.ceph.com/cbodley-2023-03-11_01:12:31-rbd-wip-coro-test-distro-default-smithi/ showed a number of failures in cli_generic.sh, rbd_mirror_stress.sh and rbd_mirror_ha.sh

**Comment by cbodley** [comment]
the 'ceph API tests' check fails to build with clang without the boost 1.81 changes:
```
src/tools/neorados.cc:68:5: error: no template named 'awaitable' in namespace 'boost::asio'
ba::awaitable<R::IOContext> lookup_pool(R::RADOS& r, const std::string& pname)
~~~~^
```

**Comment by cbodley** [comment]
> the rbd suite in https://pulpito.ceph.com/cbodley-2023-03-11_01:12:31-rbd-wip-coro-test-distro-default-smithi/ showed a number of failures in cli_generic.sh, rbd_mirror_stress.sh and rbd_mirror_ha.sh

the [--rerun](https://pulpito.ceph.com/cbodley-2023-03-11_01:12:31-rbd-wip-coro-test-distro-default-smithi/) was queued for so long that shaman had deleted its packages. i scheduled a [new rbd suite](https://pulpito.ceph.com/cbodley-2023-03-22_18:12:21-rbd-wip-coro-test-distro-default-smithi/) on [yesterday's builds](https://shaman.ceph.com/builds/ceph/wip-coro-test/3bc8694e58acdaf9a015c45c41e51ed9489d2769/)

**Comment by cbodley** [comment]
> the 'ceph API tests' check fails to build with clang without the boost 1.81 changes

@adamemerson do we need to pull in the boost 1.81 commits from https://github.com/ceph/ceph/pull/49737?

**Pull request closed without merging**